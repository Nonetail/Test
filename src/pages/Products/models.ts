export enum ProductCategory {
	ELECTRONICS = 'electronics',
	CLOTHING = 'clothing',
	FOOD = 'food',
	BOOKS = 'books',
}
