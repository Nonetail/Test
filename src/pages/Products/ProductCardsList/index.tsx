import ProductCard, { ProductCardProps } from '../../../component/ProductCard';
import { useCart } from '../../../context/CartContext';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';

interface ProductCardsListProps {
	products: Omit<ProductCardProps, 'addToCart'>[];
}

const ProductCardsList = (props: ProductCardsListProps) => {
	const { products } = props;
	const { addToCart } = useCart();
	const addToCartHandler = (id: number) => {
		addToCart(id, 1);
	};
	return (
		<Box sx={{ flexGrow: 1 }}>
			<Grid
				container
				spacing={{ xs: 2, md: 3 }}
				columns={{ xs: 4, sm: 8, md: 12 }}
			>
				{products.map((product, index) => (
					<Grid item xs={4} sm={4} md={4} key={index}>
						<ProductCard
							key={index}
							{...product}
							addToCart={addToCartHandler}
						/>
					</Grid>
				))}
			</Grid>
		</Box>
	);
};

export default ProductCardsList;
