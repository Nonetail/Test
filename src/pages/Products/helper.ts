export const getFilteredProducts = async (query?: string) => {
	const response = await fetch(`http://localhost:3000/products?${query}`);
	return await response.json();
};

export const getCategories = async () => {
	const response = await fetch(`http://localhost:3000/products/catagories`);
	return await response.json();
};
