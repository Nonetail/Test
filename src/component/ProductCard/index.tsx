import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';
import { Button } from '@mui/material';
import { ProductCategory } from '../../pages/Products/models';

export interface ProductCardProps {
	id: number;
	imgSrc: string;
	imgAltText: string;
	heading: string;
	description: string;
	category: ProductCategory;
	price: number;
	addToCart: (id: number) => void;
}

export default function ProductCard(props: ProductCardProps) {
	const {
		id,
		imgSrc,
		imgAltText,
		heading,
		description,
		category,
		price,
		addToCart,
	} = props;
	return (
		<Card sx={{ maxWidth: 345, margin: 'auto' }}>
			<CardMedia
				component='img'
				height='140'
				image={imgSrc}
				alt={imgAltText}
			/>
			<CardContent>
				<Typography gutterBottom variant='h5' component='div'>
					{heading}
				</Typography>
				<Typography variant='body2' color='text.secondary'>
					{description}
					{category}
				</Typography>
				{price}
				<Button onClick={() => addToCart(id)} color='inherit'>
					Add
				</Button>
			</CardContent>
		</Card>
	);
}
