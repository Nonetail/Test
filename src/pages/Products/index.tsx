import { useCallback, useEffect, useState } from 'react';
import ProductCardsList from './ProductCardsList';
import Cart from '../../component/Cart';
import { useCart } from '../../context/CartContext';
import SearchInput from '../../component/SearchInput';
// import debounce from 'lodash/debounce';
import Filter from '../../component/Filter';
import { ProductCategory } from './models';
import { getCategories, getFilteredProducts } from './helper';

const Products = () => {
	const [products, setProducts] = useState([]);
	const [categories, setCategories] = useState([]);
	console.log('ðŸš€ ~ Products ~ categories:', products, categories);
	const [searchTerm, setSearchTerm] = useState('');
	console.log('ðŸš€ ~ Products ~ searchTerm:', searchTerm);
	const [filterTerm, setFilterTerm] = useState('');
	const { cart } = useCart();

	// unit test
	const productCount = cart.reduce((acc, item) => acc + item.quantity, 0);

	useEffect(() => {
		Promise.all([getFilteredProducts(), getCategories()])
			.then(([products, catagories]) => {
				setProducts(products.data);
				setCategories(catagories);
			})
			.catch(e => {
				console.log('ðŸš€ ~ useEffect ~ e:', e);
			});
	}, []);

	const debounceSearch = useCallback(
		async (searchTerm: string, filterTerm: string) => {
			try {
				const response = await getFilteredProducts(
					`search=${searchTerm}&category=${filterTerm}`,
				);
				setProducts(response.data);
			} catch (error) {
				console.error('Error fetching products', error);
			}
		},
		[],
	);

	useEffect(() => {
		debounceSearch(searchTerm, filterTerm);
	}, [searchTerm, filterTerm, debounceSearch]);

	return (
		<div>
			<SearchInput
				handleSearch={event => {
					console.log(event.target.value);
					setSearchTerm(event.target.value);
				}}
			/>
			<Filter
				options={[
					{ label: 'Category 1', value: ProductCategory.ELECTRONICS },
					{ label: 'Category 2', value: ProductCategory.CLOTHING },
				]}
				handleFilter={value => setFilterTerm(value.join(','))}
			/>
			<Cart count={productCount} handleCartClick={() => {}} />
			<ProductCardsList products={products} />
		</div>
	);
};

export default Products;
