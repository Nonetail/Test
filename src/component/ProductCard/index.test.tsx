import { render } from 'react-dom';
import ProductCard from './index';
import { ProductCategory } from '../../pages/Products/models';

describe('ProductCard', () => {
	const product = {
		id: 1,
		imgSrc: 'path/to/smartphone.jpg',
		imgAltText: 'Smartphone Image',
		heading: 'Smartphone',
		description: 'High-performance smartphone with advanced features.',
		category: ProductCategory.ELECTRONICS,
		price: 499.99,
	};

	it('renders correctly', () => {
		const container = document.createElement('div'); // Create a container element
		render(<ProductCard {...product} addToCart={() => {}} />, container); // Render the ProductCard component
		expect(container).toMatchSnapshot();
	});
});
