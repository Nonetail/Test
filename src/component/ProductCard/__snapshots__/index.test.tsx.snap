// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard renders correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-yvvig0-MuiPaper-root-MuiCard-root"
  >
    <img
      alt="Smartphone Image"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
      height="140"
      src="path/to/smartphone.jpg"
    />
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
      >
        Smartphone
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
      >
        High-performance smartphone with advanced features.
        electronics
      </p>
      499.99
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-1y942vo-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add
      </button>
    </div>
  </div>
</div>
`;
